### ISSUE TITLE
[Performance] onnxruntime.InferenceSession is too slow on Ubuntu

### ISSUE THREAD

**Issue by ysq151944** [microsoft/onnxruntime#15165]
### Describe the issue

time cost of rt.InferenceSession on Ubuntu is 27.8 times to that on Windows

### To reproduce

Model Address:
https://drive.google.com/file/d/1zCpQI73w4hZ52StAZrCeOoOatD6JHtx7/view?usp=share_link
============================================================================

Code:

import time
import onnxruntime as rt

so = rt.SessionOptions()
so.inter_op_num_threads = so.intra_op_num_threads = 0
dt_st = time.perf_counter()
rt.InferenceSession("./tmp/clf.onnx", sess_options=so)
print(f'\n{so.inter_op_num_threads=}, {so.intra_op_num_threads=}'
      f'\nWindows -- tc load clf_0: {time.perf_counter() - dt_st:.5f}s')

so.inter_op_num_threads = so.intra_op_num_threads = 1
dt_st = time.perf_counter()
rt.InferenceSession(r"./tmp/clf.onnx", sess_options=so)
print(f'\n{so.inter_op_num_threads=}, {so.intra_op_num_threads=}'
      f'\nWindows -- tc load clf_1: {time.perf_counter() - dt_st:.5f}s')

============================================================================
Result:

1. 10900k + Windows10+python3.8+onnxruntime 1.14.1(CPU) + skl2onnx 1.14.0:
so.inter_op_num_threads=0, so.intra_op_num_threads=0
Windows -- tc load clf_0: 9.73551s

so.inter_op_num_threads=1, so.intra_op_num_threads=1
Windows -- tc load clf_1: 9.69827s

2. 13600k + Ubuntu22.04.0 + kernal 5.15.0-57-generic +python3.8+onnxruntime 1.14.1(CPU) + skl2onnx 1.14.0:
so.inter_op_num_threads=0, so.intra_op_num_threads=0
Ubuntu -- tc load clf_0: 272.98183s

so.inter_op_num_threads=1, so.intra_op_num_threads=1
Ubuntu -- tc load clf_1: 273.83692s

### Urgency

18 days remained for my deploy plan

### Platform

Linux

### OS Version

Ubuntu 22.04.01

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.14.1

### ONNX Runtime API

Python

### Architecture

X64

### Execution Provider

Default CPU

### Execution Provider Library Version

CPUProvider

### Model File

https://drive.google.com/file/d/1zCpQI73w4hZ52StAZrCeOoOatD6JHtx7/view?usp=share_link

I don't know what quantized model is. The onx model is converted from scikit-learn RFC via skl2onnx.convert_sklearn

### Is this a quantized model?

Unknown

**Comment by skottmckay** [comment]
The time for creating an InferenceSession is meaningless as it is the time to load and optimize the model and prepare it for execution.

What matters is the performance of running the model, and you should be doing a warmup call to `run` before measuring if you want accurate results.

**Comment by ysq151944** [comment]
> The time for creating an InferenceSession is meaningless as it is the time to load and optimize the model and prepare it for execution.
> 
> What matters is the performance of running the model, and you should be doing a warmup call to `run` before measuring if you want accurate results.

Thanks for your reply, yet I think the loading time matters in my situation:

1. the model predicts a sample every 3 minutes
2. the model should be updated every day
3. Inter-Process Communication is last thing I would try


Maybe what dosen't matter is the compiling time, not the loading time?

**Comment by skottmckay** [comment]
The expected usage is to create an InferenceSession and re-use it when you want to run the model to generate an inference, as loading the model is an expensive operation. 

i.e. create the inference session once and every 3 minutes when you need to make a prediction call `InferenceSession.run`.

If you need to update the model, create a new InferenceSession with the updated model, as the model is immutable once the InferenceSession has been initialized. 

Maybe take a look at some of the examples.

https://onnxruntime.ai/docs/get-started/with-python.html#scikit-learn-cv
https://github.com/onnx/onnx-docker/blob/master/onnx-ecosystem/inference_demos/simple_onnxruntime_inference.ipynb

**Comment by ysq151944** [comment]
> The expected usage is to create an InferenceSession and re-use it when you want to run the model to generate an inference, as loading the model is an expensive operation.
> 
> i.e. create the inference session once and every 3 minutes when you need to make a prediction call `InferenceSession.run`.
> 
> If you need to update the model, create a new InferenceSession with the updated model, as the model is immutable once the InferenceSession has been initialized.
> 
> Maybe take a look at some of the examples.
> 
> https://onnxruntime.ai/docs/get-started/with-python.html#scikit-learn-cv https://github.com/onnx/onnx-docker/blob/master/onnx-ecosystem/inference_demos/simple_onnxruntime_inference.ipynb

Thanks, I've read the doc, and I'am not worry about the inferencing duration, onnxruntime performs perfect in that part in my case(from 100ms to 16us on Ubuntu, 66us on Windows).

However, I can't "create a new InferenceSession with the updated model " unless stopping the active process or claiming new compute resource(which I would not like to do).

As you said once, "When ORT loads the model we run all our optimizers, convert to ORT internal data types, potentially do pre-packing of weights, create the allocation plan, etc.".

So, would it possible that you split the onnxruntime.InferenceSession into separate parts:
1. do all your optimization
2. let user saves the optimized model on disk
3. load optimized model directly

**Comment by xadupre** [comment]
The model contains one TreeEnsembleClassifier (you could remove the ZipMap operator: https://onnx.ai/sklearn-onnx/parameterized.html#zipmap). The optimizrs do not change the model. You can disable them. However, onnxruntime creates the tree structure every time the session is loaded. The cost should be similar to the one used to deserialize the model. This one cannot disappear and your model is 200 Mb. I don't know if protobuf behaves differently on Windows and Ubuntu. onnxruntime cannot be pickled so the loaded time cannot be avoided with onnxruntime. But it could probably with a python runtime. loading time + inference time would probably be less then 9s with a picklable python runtime you can unpickle at every call.

**Comment by ysq151944** [comment]
> The model contains one TreeEnsembleClassifier (you could remove the ZipMap operator: https://onnx.ai/sklearn-onnx/parameterized.html#zipmap). The optimizrs do not change the model. You can disable them. However, onnxruntime creates the tree structure every time the session is loaded. The cost should be similar to the one used to deserialize the model. This one cannot disappear and your model is 200 Mb. I don't know if protobuf behaves differently on Windows and Ubuntu. onnxruntime cannot be pickled so the loaded time cannot be avoided with onnxruntime. But it could probably with a python runtime. loading time + inference time would probably be less then 9s with a picklable python runtime you can unpickle at every call.

Thank you! The python runtime workes!

I not sure whether "pyhton runtime" means mlprodict.onnxrt.OnnxInference('*.onnx', runtime='python_compiled')? Anyway it works for me:

1. the loading time drops to 9.06(quicker than 11.13s on windows)
2. inference time now 2 times to onnxruntime on Ubuntu(33us / 16us)
3. on windows the inference time now 0.83 times to onnxruntime(55us / 66us)

**Comment by xadupre** [comment]
That's an option. Most of this python runtime moved to onnx.reference.ReferenceEvaluator except the c++ implementations. For the time being, in your case, OnnxInference is probably the best option.

**Comment by ysq151944** [comment]
Yes, surely OnnxInference meets all my needs. 

Thanks again! I'll close this issue.
